src/webgpu/shader/execution/evaluation_order.spec.ts

```diff
@@ -0,0 +1,480 @@
+      _result: 0,
+    },
+    {
+      name: 'SEAndMemberAccess',
+      _body: 'return set_vec4_x(&vec4_zero, 42) + vec4_zero.x;',
+      _result: 42,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('binary_logical')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#logical-expr')
+  .desc('Tests order of evaluation of logical binary expressions.')
+  .paramsSimple([
+    {
+      name: 'BothSE',
+      _body:
+        'let r = (add(&a, 1) == 3) && (add(&a, 1) == 4);' + //
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'LeftSE',
+      _body:
+        'let r = (add(&a, 1) == 3) && (a == 3);' + //
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'RightSE',
+      _body:
+        'let r = (a == 2) && (add(&a, 1) == 3);' + //
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'LeftmostSE',
+      _body:
+        'let r = (mul3_ret0(&a, &b, &c, 10) == 0) && (a == 20) && (b == 30) && (c == 40);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'RightmostSE',
+      _body:
+        'let r = (a == 2) && (b == 3) && (c == 4) && (mul3_ret0(&a, &b, &c, 10) == 0);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'MiddleSE',
+      _body:
+        'let r = (a == 2) && (b == 3) && (mul3_ret0(&a, &b, &c, 10) == 0) && (c == 40);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_And_LhsOnly',
+      _body:
+        // rhs should not execute
+        'let t = (a != 2) && (mul(&a, 10) == 20);' + //
+        'let r = (a == 2);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_And_LhsAndRhs',
+      _body:
+        // rhs should execute
+        'let t = (a == 2) && (mul(&a, 10) == 20);' + //
+        'let r = (a == 20);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_Or_LhsOnly',
+      _body:
+        // rhs should not execute
+        'let t = (a == 2) || (mul(&a, 10) == 20);' + //
+        'let r = (a == 2);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_Or_LhsAndRhs',
+      _body:
+        // rhs should execute
+        'let t = (a != 2) || (mul(&a, 10) == 20);' + //
+        'let r = (a == 20);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('binary_mixed')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#arithmetic-expr')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#logical-expr')
+  .desc('Tests order of evaluation of both arithmetic and logical binary expressions.')
+  .paramsSimple([
+    {
+      name: 'ArithAndLogical',
+      _body: 'return mul(&a, 10) - bool_to_i32(mul(&a, 10) == 200 && mul(&a, 10) == 2000);',
```

Comment:
Can we have some variants where we short circuit? e.g:
`return mul(&a, 10) - bool_to_i32(mul(&a, 10) != 200 && mul(&a, 10) == 2000);'`

Reply:
Done.